### PR TITLE
feat(meta): compute reports parallelism to meta

### LINKED ISSUE #3291: feature: Compute nodes should report parallelism (aka. cpu cores) to Meta
Currently, the meta service simply allocates 4 parallel units for each CN.

https://github.com/singularity-data/risingwave/blob/b6d24f9863dffb58de6ed69bff7faaea57f10c97/src/meta/src/cluster/mod.rs#L42

https://github.com/singularity-data/risingwave/blob/b6d24f9863dffb58de6ed69bff7faaea57f10c97/src/meta/src/cluster/mod.rs#L114-L117

A better design is to let the CN reports its number of parallel units (aka. number of CPU cores) to 
Meta service when registering itself.

### PR DESCRIPTION
**Pull request by HuaHuaY**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Let the CN report its number of parallel units. If we don't specify the number, it will report its number of cpu cores.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
close #3291

### CODE REVIEW

File: src/meta/src/cluster/mod.rs [hunk 1]
```diff
@@ -96,12 +96,13 @@ where
...
-        host_address: HostAddress,
         r#type: WorkerType,
-    ) -> Result<(WorkerNode, bool)> {
+        host_address: HostAddress,
+        worker_node_parallelism: usize,
+    ) -> Result<WorkerNode> {
         let mut core = self.core.write().await;
         match core.get_worker_by_host(host_address.clone()) {
-            Some(worker) => Ok((worker.to_protobuf(), false)),
+            Some(worker) => Ok(worker.to_protobuf()),
```
Comment (BugenZhao) [inline src/meta/src/cluster/mod.rs:105]: We may return error here.

File: src/compute/src/lib.rs [hunk 2]
```diff
@@ -73,6 +73,11 @@ pub struct ComputeNodeOpts {
     /// Enable reporting tracing information to jaeger
     #[clap(long)]
     pub enable_jaeger_tracing: bool,
+
+    /// The number of parallel units.
+    /// We will use the number of cpu cores if not specified.
+    #[clap(long, default_value = "0")]
```
Comment (BugenZhao) [inline src/compute/src/lib.rs:79]: Try this?
```suggestion
    #[clap(long, default_value_t = num_cpus::get())]
```

https://github.com/clap-rs/clap/blob/v3.2.7/examples/tutorial_derive/03_05_default_values.rs

**Comment by skyzh** [comment]
It would be better to postpone this PR. We should first implement scheduling to be based on a fixed number of parallelism (e.g. 4 parallelism), and then enable reporting parallelism from compute node.

**Comment by skyzh** [comment]
... or we can read the config from the original place `unsafe_*`. This would be consistent with before.

File: src/meta/src/cluster/mod.rs (hunk 1 above)
Comment (HuaHuaY) [reply src/meta/src/cluster/mod.rs:104]: We can discuss about this later.

**Comment by HuaHuaY** [comment]
Now we read parallelism from `risingwave.toml` instead of cli. If there is not a value, compute will report its number of cpu cores.

File: src/common/src/config.rs [hunk 3]
```diff
@@ -284,8 +284,8 @@ mod default {
         8
     }
 
-    pub fn unsafe_worker_node_parallel_degree() -> usize {
-        4
+    pub fn worker_node_parallelism() -> usize {
+        num_cpus::get()
```
Comment (skyzh) [inline src/common/src/config.rs:288]: Would this function in std lib have the same functionality as num_cpus? https://doc.rust-lang.org/stable/std/thread/fn.available_parallelism.html

**Review by skyzh — APPROVED** [review]
Rest LGTM

File: src/config/risingwave.toml [hunk 4]
```diff
@@ -6,6 +6,7 @@ heartbeat_interval_ms = 1000
 [streaming]
 checkpoint_interval_ms = 250
 in_flight_barrier_nums = 40
+worker_node_parallelism = 4
```
Comment (fuyufjh) [inline src/config/risingwave.toml:9]: Ideally, we should use different config files for CI, playground or production. Let's do it in the future.

**Review by fuyufjh — APPROVED** [review]
LGTM

File: src/meta/src/cluster/mod.rs (hunk 1 above)
Comment (yezizp2012) [reply src/meta/src/cluster/mod.rs:104]: CN will re-register after it restarts. I think we should check and update worker_node_parallelism here if CN already exist. This indicates that the CN's configuration is updated.

**Pull request merged**